{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('default')\n",
    "\n",
    "from segmenter import segment_to_p\n",
    "from browser import SPops, Pops\n",
    "\n",
    "locvars = ['Chromosome', 'start', 'end']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = 0.05/311\n",
    "key = 'btreeCmedi'\n",
    "\n",
    "spopD = {'EAS': 'East Asian', 'EUR': 'European', 'AFR':'African', 'AMR': 'Ad-Mixed American', 'SAS':'South East Asian'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{table}\n",
      "\\centering\n",
      "\\caption{Most significant directional selection hits for East Asian}\n",
      "\\begin{tabular}{lrllllp{6cm}}\n",
      "\\toprule\n",
      "Population &  Chromosome & start &   end &            p value & Average $\\hat{\\beta}$ &                                                                                                                                          genes \\\\\n",
      "\\midrule\n",
      "       JPT &           1 &  0.75 &  0.76 &  $4\\times10^{-15}$ &                 -2.27 &                                                                                                                           ACADM, RABGGTB, MSH4 \\\\\n",
      "       KHV &           2 &  0.17 &  0.18 &  $4\\times10^{-13}$ &                 -2.39 &                                                                                                                                RAD51AP2, VSNL1 \\\\\n",
      "       KHV &           2 &  1.09 &  1.10 &  $1\\times10^{-12}$ &                 -2.02 &                                                                                                        RANBP2, CCDC138, EDAR, SH3RF3, SEPTIN10 \\\\\n",
      "       CHB &          15 &  0.64 &  0.65 &  $1\\times10^{-12}$ &                 -1.79 &  CIAO2A, SNX1, SNX22, PPIB, CSNK1G1, PCLAF, TRIP4, ZNF609, OAZ2, RBPMS2, PIF1, PLEKHO2, ANKDD1A, SPG21, MTFMT, SLC51B, RASL12, KBTBD13, UBAP1L \\\\\n",
      "       CDX &           2 &  0.43 &  0.44 &  $4\\times10^{-10}$ &                 -1.89 &                                                                                            PLEKHH2, C1GALT1C1L, DYNC2LI1, ABCG5, ABCG8, LRPPRC \\\\\n",
      "       JPT &           4 &  1.43 &  1.44 &  $7\\times10^{-10}$ &                 -1.57 &                                                                                                               SMARCA5, FREM3, GYPE, GYPB, GYPA \\\\\n",
      "       CDX &           1 &  0.92 &  0.93 &  $1\\times10^{-09}$ &                 -1.39 &                                                                                     GFI1, EVI5, RPL5, DIPK1A, MTF2, TMED5, CCDC18, DR1, FNBP1L \\\\\n",
      "       JPT &           3 &  0.17 &  0.18 &  $1\\times10^{-09}$ &                 -1.15 &                                                                                                                           PLCL2, TBC1D5, SATB1 \\\\\n",
      "       KHV &           4 &  1.00 &  1.00 &  $1\\times10^{-09}$ &                 -2.42 &                                                                                                                                         DDIT4L \\\\\n",
      "       CDX &           2 &  1.97 &  1.98 &  $2\\times10^{-09}$ &                 -1.91 &                                                                                                       SF3B1, COQ10B, HSPD1, HSPE1, MOB4, RFTN2 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\\end{table}\n",
      "\n",
      "\\begin{table}\n",
      "\\centering\n",
      "\\caption{Most significant directional selection hits for European}\n",
      "\\begin{tabular}{lrllllp{6cm}}\n",
      "\\toprule\n",
      "Population &  Chromosome & start &   end &            p value & Average $\\hat{\\beta}$ &                                                                                                    genes \\\\\n",
      "\\midrule\n",
      "       CEU &           2 &  1.35 &  1.37 &  $2\\times10^{-40}$ &                 -2.31 &                                                           R3HDM1, UBXN4, LCT, MCM6, DARS1, CXCR4, THSD7B \\\\\n",
      "       GBR &          12 &  1.11 &  1.13 &  $3\\times10^{-18}$ &                 -1.50 &                 ATXN2, BRAP, ACAD10, ALDH2, MAPKAPK5, TMEM116, ERP29, NAA25, TRAFD1, RPL6, PTPN11, RPH3A \\\\\n",
      "       FIN &          15 &  0.28 &  0.29 &  $4\\times10^{-15}$ &                 -1.73 &                                                                                         GOLGA8F, GOLGA8G \\\\\n",
      "       TSI &          11 &  0.89 &  0.89 &  $5\\times10^{-13}$ &                 -1.63 &                                                                                                      TYR \\\\\n",
      "       TSI &           5 &  1.30 &  1.31 &  $4\\times10^{-11}$ &                 -1.33 &                                                                            CHSY3, HINT1, LYRM7, CDC42SE2 \\\\\n",
      "       GBR &           8 &  0.12 &  0.13 &  $5\\times10^{-11}$ &                 -1.98 &                                                                                        DEFB130A, FAM86B2 \\\\\n",
      "       FIN &           6 &  0.28 &  0.29 &  $9\\times10^{-11}$ &                 -1.35 &                                                                                              GPX5, ZBED9 \\\\\n",
      "       GBR &          15 &  0.75 &  0.75 &  $3\\times10^{-10}$ &                 -1.39 &  CLK3, EDC3, CYP1A1, CYP1A2, CSK, LMAN1L, CPLX3, ULK3, SCAMP2, MPI, FAM219B, COX5A, RPP25, SCAMP5, PPCDC \\\\\n",
      "       GBR &           6 &  0.35 &  0.35 &  $8\\times10^{-10}$ &                 -1.42 &                                                                        TCP11, SCUBE3, ZNF76, DEF6, PPARD \\\\\n",
      "       CEU &           1 &  1.00 &  1.00 &  $1\\times10^{-09}$ &                 -1.87 &                                                                                                      AGL \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\\end{table}\n",
      "\n",
      "\\begin{table}\n",
      "\\centering\n",
      "\\caption{Most significant directional selection hits for African}\n",
      "\\begin{tabular}{lrllllp{6cm}}\n",
      "\\toprule\n",
      "Population &  Chromosome & start &   end &            p value & Average $\\hat{\\beta}$ &                                                                 genes \\\\\n",
      "\\midrule\n",
      "       ESN &          19 &  0.43 &  0.44 &  $4\\times10^{-17}$ &                 -2.19 &  PSG5, PSG4, PSG9, CD177, TEX101, LYPD3, PHLDB3, ETHE1, ZNF575, XRCC1 \\\\\n",
      "       ESN &           6 &  0.53 &  0.53 &  $5\\times10^{-16}$ &                 -2.74 &                                                               TMEM14A \\\\\n",
      "       GWD &           6 &  0.59 &  0.62 &  $9\\times10^{-14}$ &                 -2.25 &                                                               KHDRBS2 \\\\\n",
      "       LWK &           8 &  0.44 &  0.47 &  $1\\times10^{-12}$ &                 -1.60 &                                                                 SPIDR \\\\\n",
      "       ESN &          11 &  0.49 &  0.49 &  $2\\times10^{-11}$ &                 -1.60 &                                                      TRIM49B, TRIM64C \\\\\n",
      "       YRI &           1 &  1.62 &  1.62 &  $2\\times10^{-11}$ &                 -1.80 &                                                          DUSP12, ATF6 \\\\\n",
      "       ESN &           1 &  1.54 &  1.54 &  $7\\times10^{-11}$ &                 -1.73 &                   CRTC2, SLC39A1, CREB3L4, JTB, RAB13, RPS27, NUP210L \\\\\n",
      "       ESN &           3 &  1.25 &  1.26 &  $2\\times10^{-09}$ &                 -1.61 &                                                         SNX4, OSBPL11 \\\\\n",
      "       ASW &          13 &  0.52 &  0.52 &  $4\\times10^{-09}$ &                 -1.43 &                                                                  NEK3 \\\\\n",
      "       YRI &          19 &  0.39 &  0.39 &  $4\\times10^{-09}$ &                 -1.68 &                    LGALS7, LGALS7B, LGALS4, ECH1, HNRNPL, RINL, SIRT2 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\\end{table}\n",
      "\n",
      "\\begin{table}\n",
      "\\centering\n",
      "\\caption{Most significant directional selection hits for Ad-Mixed American}\n",
      "\\begin{tabular}{lrllllp{6cm}}\n",
      "\\toprule\n",
      "Population &  Chromosome & start &   end &            p value & Average $\\hat{\\beta}$ &                                                                        genes \\\\\n",
      "\\midrule\n",
      "       PUR &          22 &  0.29 &  0.29 &  $1\\times10^{-07}$ &                 -1.87 &                          EMID1, RHBDD3, EWSR1, GAS2L1, RASL10A, AP1B1, RFPL1 \\\\\n",
      "       CLM &           6 &  1.35 &  1.35 &  $2\\times10^{-07}$ &                 -2.18 &                                                                         AHI1 \\\\\n",
      "       CLM &           8 &  0.44 &  0.48 &  $2\\times10^{-07}$ &                 -1.54 &                                                                        SPIDR \\\\\n",
      "       PUR &           3 &  1.29 &  1.30 &  $4\\times10^{-07}$ &                 -1.91 &                                     EFCAB12, MBD4, IFT122, RHO, H1-8, PLXND1 \\\\\n",
      "       PUR &          17 &  0.37 &  0.37 &  $5\\times10^{-07}$ &                 -2.39 &                                                                        ACACA \\\\\n",
      "       MXL &           2 &  0.24 &  0.25 &  $8\\times10^{-07}$ &                 -1.99 &                                                                        NCOA1 \\\\\n",
      "       PUR &          12 &  0.44 &  0.48 &  $1\\times10^{-06}$ &                 -0.94 &  NELL2, DBX2, ANO6, ARID2, SCAF11, SLC38A1, SLC38A2, SLC38A4, AMIGO2, PCED1B \\\\\n",
      "       MXL &           8 &  0.86 &  0.87 &  $1\\times10^{-06}$ &                 -1.71 &                   PSKH2, ATP6V0D2, SLC7A13, WWP1, RMDN1, CPNE3, CNGB3, CNBD1 \\\\\n",
      "       PUR &          11 &  0.13 &  0.13 &  $1\\times10^{-06}$ &                 -1.92 &                                                                        TEAD1 \\\\\n",
      "       PUR &           7 &  1.29 &  1.29 &  $2\\times10^{-06}$ &                 -1.84 &                                              ATP6V1F, ATP6V1FNB, IRF5, TNPO3 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\\end{table}\n",
      "\n",
      "\\begin{table}\n",
      "\\centering\n",
      "\\caption{Most significant directional selection hits for South East Asian}\n",
      "\\begin{tabular}{lrllllp{6cm}}\n",
      "\\toprule\n",
      "Population &  Chromosome & start &   end &            p value & Average $\\hat{\\beta}$ &                                                                                                                                                                                                                                                                                                                            genes \\\\\n",
      "\\midrule\n",
      "       BEB &          16 &  0.34 &  0.47 &  $3\\times10^{-17}$ &                 -1.45 &                                                                                                                                                                                                                                                                                       SHCBP1, VPS35, ORC6, MYLK3, C16orf87, GPT2 \\\\\n",
      "       GIH &          15 &  0.48 &  0.49 &  $3\\times10^{-12}$ &                 -1.69 &                                                                                                                                                                                                                                                                                                                        DUT, FBN1 \\\\\n",
      "       STU &          17 &  0.44 &  0.44 &  $3\\times10^{-12}$ &                 -1.77 &                                                                                                                                                                                                                                                      HROB, ASB16, TMUB2, ATXN7L3, UBTF, SLC4A1, RUNDC3A, SLC25A39, GRN, FAM171A2 \\\\\n",
      "       GIH &           3 &  0.89 &  0.94 &  $8\\times10^{-12}$ &                 -0.84 &                                                                                                                                                                                                                                                                                                                     EPHA3, PROS1 \\\\\n",
      "       GIH &           2 &  0.97 &  0.97 &  $1\\times10^{-11}$ &                 -1.73 &                                                                                                                                                                                                                                                                                                                           FAHD2B \\\\\n",
      "       STU &          16 &  0.30 &  0.31 &  $4\\times10^{-11}$ &                 -1.23 &  SEPHS2, ITGAL, ZNF768, ZNF747, ZNF764, ZNF688, ZNF785, ZNF689, PRR14, FBRS, SRCAP, TMEM265, PHKG2, CCDC189, RNF40, ZNF629, BCL7C, CTF1, FBXL19, ORAI3, SETD1A, HSD3B7, STX1B, STX4, ZNF668, ZNF646, PRSS53, VKORC1, BCKDK, KAT8, PRSS8, PRSS36, FUS, PYCARD, TRIM72, PYDC1, ITGAM, ITGAX, ITGAD, COX6A2, ZNF843, ARMC5, TGFB1I1 \\\\\n",
      "       PJL &           2 &  2.23 &  2.23 &  $5\\times10^{-11}$ &                 -1.60 &                                                                                                                                                                                                                                                                                                                            FARSB \\\\\n",
      "       PJL &           2 &  0.92 &  0.95 &  $6\\times10^{-11}$ &                 -1.63 &                                                                                                                                                                                                                                                                                   TEKT4, MAL, MRPS5, ZNF514, ZNF2, PROM2, KCNIP3 \\\\\n",
      "       STU &          10 &  0.93 &  0.94 &  $1\\times10^{-10}$ &                 -1.20 &                                                                                                                                                                                                                                                                        CEP55, FFAR4, RBP4, PDE6C, FRA10AC1, LGI1, SLC35G1, PLCE1 \\\\\n",
      "       STU &          19 &  0.36 &  0.36 &  $4\\times10^{-10}$ &                 -1.90 &                                                                                                                                                                                                                                                                                                                            ZFP82 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\\end{table}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for spop in ['EAS', 'EUR', 'AFR', 'AMR', 'SAS']:\n",
    "    dfs = []\n",
    "    for i in range(26):\n",
    "        if SPops[i] == spop:\n",
    "            df = pd.read_csv('gene_scores/'+str(i)+'_'+key+'.csv')\n",
    "            df.insert(0, 'Population', Pops[i][:3])\n",
    "            dfs.append(df)\n",
    "\n",
    "    dfs = pd.concat(dfs)   \n",
    "    \n",
    "    dx = dfs.sort_values('pbtreeCmedi').reset_index(drop = True).iloc[:100,:]\n",
    "\n",
    "    cur_genes = []\n",
    "\n",
    "    i = 0\n",
    "    dels = []\n",
    "\n",
    "    for i in range(dx.shape[0]):\n",
    "        try:\n",
    "            curg = dx.iloc[i]['genes'].replace(',','').split(' ')\n",
    "        except:\n",
    "            dels.append(i)\n",
    "        if any([i in cur_genes for i in curg]):\n",
    "            dels.append(i)\n",
    "\n",
    "        cur_genes=cur_genes+curg\n",
    "\n",
    "    dx = dx.drop(dels).iloc[:10,:].reset_index(drop=True)\n",
    "    \n",
    "    dx = dx.drop(columns = ['sbtreeCmedi'])\n",
    "    \n",
    "    dx['start'] = dx['start'].apply(lambda x: '{:0.2f}'.format(x/1e8))\n",
    "    dx['end'] = dx['end'].apply(lambda x: '{:0.2f}'.format(x/1e8))\n",
    "    \n",
    "    dx['p'+key] = dx['p'+key].apply(lambda x: r'$'+'{:1.0}'.format(x).replace('e-', '\\times10^{-')+'}$')\n",
    "    dx['btreeCmedi'] = dx['btreeCmedi'].apply(lambda x: '{:0.2f}'.format(x))\n",
    "    dz = dx.rename(columns = {'btreeCmedi':r'Average $\\hat{\\beta}$', 'pbtreeCmedi':'p value'})\n",
    "    pd.set_option('max_colwidth', 400)\n",
    "    print(dz.to_latex(index=False, escape = False, caption = 'Most significant directional selection hits for '+spopD[spop], column_format = 'lrllllp{6cm}'))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{table}\n",
      "\\centering\n",
      "\\caption{Most significant directional selection hits for 1000 Genomes Project Populations}\n",
      "\\begin{tabular}{lrllllp{6cm}}\n",
      "\\toprule\n",
      "Population &  Chromosome & start &   end &            p value & Average $\\hat{\\beta}$ &                                                                                     genes \\\\\n",
      "\\midrule\n",
      "       CEU &           2 &  1.35 &  1.37 &  $2\\times10^{-40}$ &                 -2.31 &                                            R3HDM1, UBXN4, LCT, MCM6, DARS1, CXCR4, THSD7B \\\\\n",
      "       GBR &          12 &  1.11 &  1.13 &  $3\\times10^{-18}$ &                 -1.50 &  ATXN2, BRAP, ACAD10, ALDH2, MAPKAPK5, TMEM116, ERP29, NAA25, TRAFD1, RPL6, PTPN11, RPH3A \\\\\n",
      "       BEB &          16 &  0.34 &  0.47 &  $3\\times10^{-17}$ &                 -1.45 &                                                SHCBP1, VPS35, ORC6, MYLK3, C16orf87, GPT2 \\\\\n",
      "       ESN &          19 &  0.43 &  0.44 &  $4\\times10^{-17}$ &                 -2.19 &                      PSG5, PSG4, PSG9, CD177, TEX101, LYPD3, PHLDB3, ETHE1, ZNF575, XRCC1 \\\\\n",
      "       ESN &           6 &  0.53 &  0.53 &  $5\\times10^{-16}$ &                 -2.74 &                                                                                   TMEM14A \\\\\n",
      "       JPT &           1 &  0.75 &  0.76 &  $4\\times10^{-15}$ &                 -2.27 &                                                                      ACADM, RABGGTB, MSH4 \\\\\n",
      "       FIN &          15 &  0.28 &  0.29 &  $4\\times10^{-15}$ &                 -1.73 &                                                                          GOLGA8F, GOLGA8G \\\\\n",
      "       GWD &           6 &  0.59 &  0.62 &  $9\\times10^{-14}$ &                 -2.25 &                                                                                   KHDRBS2 \\\\\n",
      "       KHV &           2 &  0.17 &  0.18 &  $4\\times10^{-13}$ &                 -2.39 &                                                                           RAD51AP2, VSNL1 \\\\\n",
      "       TSI &          11 &  0.89 &  0.89 &  $5\\times10^{-13}$ &                 -1.63 &                                                                                       TYR \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\\end{table}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "key = 'btreeCmedi'\n",
    "\n",
    "dfs = []\n",
    "for i in range(26):\n",
    "    df = pd.read_csv('gene_scores/'+str(i)+'_'+key+'.csv')\n",
    "    df.insert(0, 'Population', Pops[i][:3])\n",
    "    dfs.append(df)\n",
    "\n",
    "dfs = pd.concat(dfs)   \n",
    "\n",
    "dx = dfs.sort_values('p'+key).reset_index(drop = True).iloc[:100,:]\n",
    "\n",
    "cur_genes = []\n",
    "\n",
    "i = 0\n",
    "dels = []\n",
    "\n",
    "for i in range(dx.shape[0]):\n",
    "    try:\n",
    "        curg = dx.iloc[i]['genes'].replace(',','').split(' ')\n",
    "    except:\n",
    "        dels.append(i)\n",
    "    if any([i in cur_genes for i in curg]):\n",
    "        dels.append(i)\n",
    "\n",
    "    cur_genes=cur_genes+curg\n",
    "\n",
    "dx = dx.drop(dels).iloc[:10,:].reset_index(drop=True)\n",
    "\n",
    "dx = dx.drop(columns = ['s'+key])\n",
    "\n",
    "dx['start'] = dx['start'].apply(lambda x: '{:0.2f}'.format(x/1e8))\n",
    "dx['end'] = dx['end'].apply(lambda x: '{:0.2f}'.format(x/1e8))\n",
    "\n",
    "dx['p'+key] = dx['p'+key].apply(lambda x: r'$'+'{:1.0}'.format(x).replace('e-', '\\times10^{-')+'}$')\n",
    "dx[key] = dx[key].apply(lambda x: '{:0.2f}'.format(x))\n",
    "dz = dx.rename(columns = {key:r'Average $\\hat{\\beta}$', 'p'+key:'p value'})\n",
    "pd.set_option('max_colwidth', 400)\n",
    "print(dz.to_latex(index=False, escape = False, caption = 'Most significant directional selection hits for 1000 Genomes Project Populations', column_format = 'lrllllp{6cm}'))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{table}\n",
      "\\centering\n",
      "\\caption{Most significant balancing selection hits for 1000 Genomes Project Popultions}\n",
      "\\begin{tabular}{lrllllp{4cm}}\n",
      "\\toprule\n",
      "Population &  Chromosome &  start &    end &      p value & Average $\\hat{\\beta}$ &                                                                                                     genes \\\\\n",
      "\\midrule\n",
      "       ASW &           6 &  0.322 &  0.327 &  $<10^{-38}$ &                  1.33 &  RNF5, AGER, PBX2, GPSM3, NOTCH4, TSBP1, BTNL2, HLA-DRA, HLA-DRB5, HLA-DRB1, HLA-DQA1, HLA-DQB1, HLA-DQA2 \\\\\n",
      "       YRI &           4 &  0.989 &  0.991 &  $<10^{-38}$ &                  4.71 &                                                                                                    METAP1 \\\\\n",
      "       IBS &           1 &  1.529 &  1.532 &  $<10^{-38}$ &                  6.43 &                              SPRR4, SPRR1A, SPRR3, SPRR1B, SPRR2D, SPRR2A, SPRR2B, SPRR2E, SPRR2F, SPRR2G \\\\\n",
      "       ESN &           4 &  0.854 &  0.855 &  $<10^{-38}$ &                  5.78 &                                                                                                  ARHGAP24 \\\\\n",
      "       ASW &           5 &  1.355 &  1.355 &  $<10^{-38}$ &                  6.61 &                                                                                                   NEUROG1 \\\\\n",
      "       ACB &          11 &  0.051 &  0.051 &  $<10^{-38}$ &                  5.16 &                                                                                                    OR52E1 \\\\\n",
      "       ASW &           6 &  0.330 &  0.331 &  $<10^{-38}$ &                  4.98 &                                                                                        HLA-DPA1, HLA-DPB1 \\\\\n",
      "       IBS &          13 &  0.368 &  0.368 &  $<10^{-38}$ &                  6.59 &                                                                                                     RFXAP \\\\\n",
      "       STU &           1 &  2.315 &  2.315 &  $<10^{-38}$ &                  5.87 &                                                                                                     TSNAX \\\\\n",
      "       STU &          20 &  0.235 &  0.236 &  $<10^{-38}$ &                  5.76 &                                                                                                     CST9L \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\\end{table}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "key = 'bsfs'\n",
    "\n",
    "dfs = []\n",
    "for i in range(26):\n",
    "    df = pd.read_csv('gene_scores/'+str(i)+'_'+key+'.csv')\n",
    "    df.insert(0, 'Population', Pops[i][:3])\n",
    "    dfs.append(df)\n",
    "\n",
    "dfs = pd.concat(dfs)   \n",
    "\n",
    "dx = dfs.sort_values('p'+key, ascending = False).reset_index(drop = True).iloc[:100,:]\n",
    "\n",
    "cur_genes = []\n",
    "\n",
    "i = 0\n",
    "dels = []\n",
    "\n",
    "for i in range(dx.shape[0]):\n",
    "    try:\n",
    "        curg = dx.iloc[i]['genes'].replace(',','').split(' ')\n",
    "    except:\n",
    "        dels.append(i)\n",
    "    if any([i in cur_genes for i in curg]):\n",
    "        dels.append(i)\n",
    "\n",
    "    cur_genes=cur_genes+curg\n",
    "\n",
    "dx = dx.drop(dels).iloc[:10,:].reset_index(drop=True)\n",
    "\n",
    "dx = dx.drop(columns = ['s'+key])\n",
    "\n",
    "dx['start'] = dx['start'].apply(lambda x: '{:0.3f}'.format(x/1e8))\n",
    "dx['end'] = dx['end'].apply(lambda x: '{:0.3f}'.format(x/1e8))\n",
    "\n",
    "#dx['p'+key] = dx['p'+key].apply(lambda x: r'$'+'{:0.1e}'.format(x).replace('e-', '\\times10^{-')+'}$')\n",
    "dx['p'+key] = r'$<10^{-38}$'\n",
    "dx[key] = dx[key].apply(lambda x: '{:0.2f}'.format(x))\n",
    "dz = dx.rename(columns = {key:r'Average $\\hat{\\beta}$', 'p'+key:'p value'})\n",
    "pd.set_option('max_colwidth', 400)\n",
    "print(dz.to_latex(index=False, escape = False, caption = 'Most significant balancing selection hits for 1000 Genomes Project Popultions', column_format = 'lrllllp{4cm}'))  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
